### PR TITLE
Refactor device structures

### PR DESCRIPTION
**Pull request by tbursztyka**
Part of #22941

Original struct device was only meant for actual devices. But its infrastructure has been then used also for non-hardware based boot-time critical piece of software. Though this was a simple solution, it brought the device concept to non-device systems. 

So refactoring all of this in order to get rid of struct device from these systems, as well as clarifying the struct device as well. 

This leads to some bytes saved from both ROM and RAM.

### CODE REVIEW

**Comment by tbursztyka** [comment]
@nashif  no idea why these timeouts are happening (on msp2_an385 and msp2_an521). It work just fine on my machine. Any clue?

File: include/init.h
Comment (mbolivar-nordic) [inline include/init.h]: Seems like this could be moved into device.h entirely, if it is really device specific. Any point to keep it here?

File: include/init.h
Comment (mbolivar-nordic) [inline include/init.h]: I am still not sure why PRE_KERNEL_1 and PRE_KERNEL_2 are meaningful. They used to be necessary, but now that we have proper dependency tracking in devicetree they do not seem needed. Further, their definitions here are about dependencies, but they don't really seem meaningful to me, since dependencies are a bit nuanced even with in the SoC: e.g. peripheral devices can depend on clock devices, etc.

Other than keeping legacy code working (which is not an objective of this effort), is there a reason to keep these levels? They seem to me to do more harm than good, but I don't know all the use cases, of course.

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks a lot for sharing this.

I have some questions about the way system initialization is changing. For now I've only thought through the device parts, but I will think a bit more about dependencies that software services have on devices and how that could fit here.

File: include/init.h
Comment (tbursztyka) [reply include/init.h]: Because init_entry stuff is not device specific, actually. It's "generic". Note that, if required, in future we could have a union {} around the dev, and a marker about the type on the init. (Though I don't any requirement for that yet. But it's possible)

Comment (mbolivar-nordic) [reply include/init.h]: Then why not just have this be a `void *arg` and be done with it?

File: include/init.h
Comment (tbursztyka) [reply include/init.h]: This PR is not about changing the levels, so no harm here.

Anyway, note that each level have priority only between 0 to 99. That limits a bit the sorting possibility (devices ordering is made on level, then priority and finally their name). That was basically the reason of splitting the pre kernel level in 2, historically, if I remember well.

**Comment by tbursztyka** [comment]
I made a trivial mistake somewhere and can't put the finger on it. (mps2_* tests/samples don't run, could be arm_mps2_init() that is not ran or not at the order it should...)

File: include/init.h
Comment (tbursztyka) [reply include/init.h]: I don't want to do casting on such low level parts. First because it will not give you any gain from memory point of view and is less readable. Plus, next phase (with the device status reporting stuff) will be to "const-ify" all struct device everywhere, with all that implies.

File: include/init.h
Comment (mbolivar-nordic) [reply include/init.h]: > This PR is not about changing the levels, so no harm here.

Are you open to the discussion of removing those levels later then? I ask because this PR seems to be specifically about refactoring initialization, and that's what the levels are part of.

Comment (tbursztyka) [reply include/init.h]: No it's refactoring device structures only. Not levels, not the initialization (it's exactly the same process, I did not change anything there).

I would not change anything about the level yet. For the good reason that it is not yet a major issue, compared to the rest. level/prio dts generation is imo of a lesser priority compared to dts device instance generation (because if that's made right, then level/prio generation will be super trivial to do)

Let's proceed step by step (and make sure people go along each of these changes)

Comment (nashif) [reply include/init.h]: > Are you open to the discussion of removing those levels later then?

Why do you want to remove them? those are not exclusive to device initialization or setup. A basic feature those levels provide is giving the developer the possibility to run code at any step of the booting process, sometimes you have to do HW initialization that is not managed by a driver or DTS and you need that at early stage. Removing those level will break many application and a useful interface.

Driver initialization priorities need to be improved and should utilise DTS for that, however, this is orthogonal and unrelated IMO. The driver priority happens on top of the existing scheme, does not replace it.

**Comment by tbursztyka** [comment]
definitely too many reviewers due to the device attribute access patch, but anyway...

**Comment by tbursztyka** [comment]
@andrewboie Should Z_SYSCALL_SPECIFIC_DRIVER be removed as it is unused? (unless I missed something). Not sure the modification I did makes that macro relevant anymore anyway.

**Comment by pabigot** [comment]
> @andrewboie Should Z_SYSCALL_SPECIFIC_DRIVER be removed as it is unused?

IMO yes.  It was added in #12056 in response to my #11993 but I never got information on how it was intended to be used, and we have an alternative solution from #18186 that handles current needs.

File: include/init.h
Comment (mbolivar-nordic) [reply include/init.h]: Can you point to a definition for _1 vs. _2 that isn't in terms of device dependencies? Right now that's how they're defined. That's a hack. We have real dependencies from DT now.

File: include/device.h
Comment (nordic-krch) [inline include/device.h]: static const?

File: include/device.h
Comment (nordic-krch) [inline include/device.h]: const?

File: kernel/device.c
Comment (nordic-krch) [inline kernel/device.c]: because of that dev must be in RAM. Any suggestions how to avoid that?

File: include/device.h
Comment (tbursztyka) [reply include/device.h]: see my answer below

File: kernel/device.c
Comment (tbursztyka) [reply kernel/device.c]: Read #22941 carefully (specially points 7, 4 and 5). 

It's already being tackled.

File: include/init.h
Comment (nashif) [reply include/init.h]: This is just a bug in the documentation, nobody bothered to change this and was long there and before device tree. So the fact it talks about devices does not mean this is 1:1 match with devicetree. Right now this is used for drivers, services and hardware initialization of different sorts that is not managed by DT, I am sure some of this can be optimized to be devicetree "devices" but we should not limit this to device tree, there is a need for this in the tree currently and there was always a need, the reason it is there.

I suggest the following change...

```diff
--- a/include/device.h
+++ b/include/device.h
@@ -57,25 +57,31 @@ extern "C" {
  * Must be one of the following symbols, which are listed in the order
  * they are performed by the kernel:
  * \n
- * \li PRE_KERNEL_1: Used for devices that have no dependencies, such as those
- * that rely solely on hardware present in the processor/SOC. These devices
+ * \li PRE_KERNEL_1: Used for units that have no dependencies, such as those
+ * that rely solely on hardware present in the processor/SOC. These units
  * cannot use any kernel services during configuration, since they are not
  * yet available.
  * \n
- * \li PRE_KERNEL_2: Used for devices that rely on the initialization of devices
- * initialized as part of the PRE_KERNEL_1 level. These devices cannot use any
- * kernel services during configuration, since they are not yet available.
+ * \li PRE_KERNEL_2: Used for units that rely on the initialization of other
+ * units initialized as part of the PRE_KERNEL_1 level. These units cannot
+ * use any kernel services during configuration, since they are not yet
+ * available.
  * \n
- * \li POST_KERNEL: Used for devices that require kernel services during
+ * \li POST_KERNEL: Used for units that require kernel services during
  * configuration.
  * \n
- * \li POST_KERNEL_SMP: Used for devices that require kernel services during
+ * \li POST_KERNEL_SMP: Used for units that require kernel services during
  * configuration after SMP initialization.
  * \n
  * \li APPLICATION: Used for application components (i.e. non-kernel components)
- * that need automatic configuration. These devices can use all services
+ * that need automatic configuration. These units can use all services
  * provided by the kernel during configuration.
  *
+ * Units are the objects that Zephyr knows how to manage. They represent system
+ * resources that can be initialised and executed based on the level and
+ * priority defined for each unit. Units in some ways are similar to services
+ * or jobs in other init systems and additionally include device drivers.
+ *
  * @param prio The initialization priority of the device, relative to
  * other devices of the same initialization level. Specified as an integer
  * value in the range 0 to 99; lower values indicate earlier initialization.
```

File: include/init.h
Comment (pabigot) [reply include/init.h]: I get the point about "less readable", but casting a `struct device *` to a `struct notdevice *` violates MISRA 11.3, which is a required constraint, compared to 11.4 which only advises against casting a `void *` to a `struct device *`.  So maybe it really is better as:
```
struct device_init_entry {
    int (*init)(struct void *p);
    void *p;
};
```
But I'm not sure if the future constifying makes that irrelevant.

File: include/init.h
Comment (andrewboie) [reply include/init.h]: We have system services via SYS_INIT() which get invoked from PRE_KERNEL phases, and I fail to see how removing this improves the kernel in a significant way.

Comment (tbursztyka) [reply include/init.h]: @anas I did change the doc already, could you check? I did it before you proposal, basically I used "initialization objects" instead of "units".

File: include/device.h
Comment (wentongwu) [inline include/device.h:257]: per #22941, what's the plan for the device dependency here? I mean the APIs you want to expose to other sub-systems, AFAIK Linux will put `struct device *parent` here to represent the dependency, but the dependency building process belongs to DTS. I'm considering the APIs of PM about how to utilize these dependencies to power management system, it will be good to know the APIs of device dependency.

Comment (tbursztyka) [reply include/device.h:257]: It is not part of this RFC. See #22941 for the planned changes.

File: include/device.h
Comment (pabigot) [inline include/device.h:259]: This should be renamed to just `pm_control`; that it appears within a `device` structure should be enough to disambiguate.  Also this makes it consistent with the existing `pm` field in the same structure.

File: include/init.h
Comment (pabigot) [inline include/init.h:44]: The review version:  We need to decide, and express clearly, whether there can be init functions that take a non-null argument that points to something other than a device.  If so, then the parameter declaration should be `void * arg` to avoid forced MISRA violations.

If the null/non-null status of the `dev` field is sufficient to distinguish device initialization from all other initializations then the types don't need to change, but the documentation should be "~Can~ Shall be NULL if and only if this....".

File: doc/reference/usermode/syscalls.rst
Comment (pabigot) [inline doc/reference/usermode/syscalls.rst:295]: As I noted earlier added in #12056 in response to my #11993 but I never got information on how it was intended to be used, e.g. examples/tests (@andrewboie).

We have a solution from #18186 that defines how to add device-specific API extensions, but it does not address supporting the extensions in user-mode.  I will have to update #17631 to implement system call checks correctly for the extended API.  I will be looking at this very soon, but the solution may use this macro.

Also this has been present in releases including 1.14, so it may be used in out of tree drivers and so arguably cannot be removed without deprecation.

**Review by pabigot — COMMENTED** [review]
Just comments for now.  We will have to clarify the init API, and whether the syscall macro can be removed.

I believe the transformations related to moving `struct deviceconfig` into `struct device` may be within the scope of what Coccinelle can do, which is more reliable than a global sed operation and also would support converting out-of-tree drivers.

File: include/device.h
Comment (tbursztyka) [reply include/device.h:259]: I am not going to touch PM stuff in this refactoring. So no name change here. But I agree this name is wrong, as well as first parameter (it should be struct device *dev, and not a name that is the same as the struct itself)

File: include/init.h
Comment (tbursztyka) [reply include/init.h:44]: The question we need to ask ourselves here: are we going to initialize objects - other than struct device - that could exist as n instances for the same initialization function?

Comment (pabigot) [reply include/init.h:44]: Agreed.  An outdated example might be all the pimux-based init functions: rather than hard-coding them into a single function sprinkled with `#ifdef` regions, the settings might be recorded in per-instance tables, and a SOC-specific init function used to process the settings for each instance that is actually linked in.

I *think* I could accept an argument that any case where an init function might be shared among multiple clients could be handled by a level of indirection: register a wrapper function with no arguments, which invokes a subordinate function on each of the objects it's supposed to affect.  This might waste space when multiple wrappers are needed to enforce dependency order.

File: doc/reference/usermode/syscalls.rst
Comment (tbursztyka) [reply doc/reference/usermode/syscalls.rst:295]: I can deprecate it but it will need to verify against api and not init function anymore then. (as it was before I removed it entirely)

Comment (pabigot) [reply doc/reference/usermode/syscalls.rst:295]: If this must be removed it'll be because it can't be made to work, but the problem it was intended to help with is still present.   Based on research summarized in #24539 I propose this PR just switch it to checking the API instead of the init pointer, and not deprecate it.

Comment (tbursztyka) [reply doc/reference/usermode/syscalls.rst:295]: Ok, back to API check then.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
NACK on removing Z_SYSCALL_SPECIFIC_DRIVER without some equivalent being proposed.

**Comment by andrewboie** [comment]
BTW It's not a hard requirement that init_fn be used, we can change the API if needed. there needs to be some pointer value passed to Z_SYSCALL_SPECIFIC_DRIVER() which can be looked up, somewhere in the `struct device`, for a match. Currently this is the init_fn but it can be something else, if the init_fn isn't stored within the device struct any more.

Having said that, from looking at the definition of `struct init_entry` you have a device pointer stored there anyway...so why not remove the `init` pointer from `struct init_entry` and just look it up in the device instead?

**Comment by tbursztyka** [comment]
@andrewboie I reverted back to changing Z_SYSCALL_SPECIFIC_DRIVER using API structure pointer instead of init_fn, rather than removing it.

File: include/init.h
Comment (pabigot) [inline include/init.h:41]: missing end of sentence.

File: include/init.h
Comment (pabigot) [reply include/init.h:45]: I propose that this issue is resolved by the previous comment block and documentation that a null device pointer field/argument indicates that the init function does not take any arguments.

**Review by pabigot — CHANGES_REQUESTED** [review]
One minor documentation fix then I'm ready to approve this pending shippable resolution.  The checkpatch diagnostics are IMO inappropriate.

**Comment by tbursztyka** [comment]
@andrewboie 
> Having said that, from looking at the definition of struct init_entry you have a device pointer stored there anyway...so why not remove the init pointer from struct init_entry and just look it up in the device instead?

I am confused about this comment, that's reverting the idea of this PR. Or did I miss something?

**Comment by carlescufi** [comment]
@andrewboie @tbursztyka is this PR a candidate for tomorrow's dev-review meeting to discuss any pending disagreements and try to get it in for 2.3?

**Comment by tbursztyka** [comment]
@carlescufi sure why not.

**Comment by carlescufi** [comment]
@mbolivar-nordic and @nordic-krch I'd like to get this into 2.3. Could you re-review and state your opinion about the readiness of the PR?

File: include/device.h
Comment (nordic-krch) [inline include/device.h:141]: done this way it adds code duplication and reduces readability. What about something like this: 
https://gist.github.com/nordic-krch/d299a388124f757c7933f0911018eece

It is using `IF_ENABLED(flag, (code)) macro to inject code only is flag is enabled. It simplifies that code and in the example i tried did not change ram usage.

File: include/device.h
Comment (nordic-krch) [inline include/device.h:98]: i don't fully get what is the difference between `DEVICE_AND_API_INIT` and `DEVICE_DEFINE`. Is it only pm support in `DEVICE_DEFINE`? If so then i naming is confusing. I would prefer to see one macro `DEVICE_DEFINE` if device has no power management then NULL is provided for `pm_control_fn`

File: kernel/device.c
Comment (nordic-krch) [inline kernel/device.c:45]: could be const?

Comment (tbursztyka) [reply kernel/device.c:45]: Indeed. Actually it's missing in various places.

File: include/device.h
Comment (tbursztyka) [reply include/device.h:98]: Yes, these macros would require some rework as well. 
Basically, there should be only one macro: DEVICE_DEFINE(), so it would follow the same naming as the kernel. Let's see this in further PR.

File: include/device.h
Comment (tbursztyka) [reply include/device.h:141]: Well, that was not the point of this PR to change the macros and how they are declared. Your proposal sounds totally relevant to me, could you post a PR?

File: include/device.h
Comment (nordic-krch) [reply include/device.h:98]: sorry, i didn't not notice that you didn't change those macros. it is messy indeed.

**Review by tyrantlizard77 — APPROVED** [review]
The syscall validation now takes an API pointer, this is fine w/me.

**Review by nashif — APPROVED** [review]

**Review by pabigot — APPROVED** [review]

File: include/init.h
Comment (mnkp) [inline include/init.h:56]: We're introducing a new `INIT_` namespace for the defines. If I correctly remember various discussions we want to use namespaces sparingly. Since this macro should not be used directly we could prefix it with Z_ as in `Z_INIT_ENTRY_DEFINE`, or maybe `SYS_`?

Comment (tbursztyka) [reply include/init.h:56]: Good point, I guess Z_ would be the one

**Review by nordic-krch — APPROVED** [review]

**Comment by pabigot** [comment]
@tbursztyka https://github.com/pabigot/zephyr/commits/nordic/pr/23407 has a commit that fixes a few missed conversions; please cherry-pick and squash it.

**Review by pabigot — CHANGES_REQUESTED** [review]
temporary block until additional fixes in https://github.com/pabigot/zephyr/commits/nordic/pr/23407 have been assessed.

**Comment by tbursztyka** [comment]
> @tbursztyka https://github.com/pabigot/zephyr/commits/nordic/pr/23407 has a commit that fixes a few missed conversions; please cherry-pick and squash it.

Would you mind sending a patch adding the cocci file you used? I think it's worth adding it into scripts/coccinelle.

**Comment by pabigot** [comment]
> Would you mind sending a patch adding the cocci file you used? I think it's worth adding it into scripts/coccinelle.

I can do that, but....

Zephyr's coccinelle infrastructure follows Linux's in expecting the semantic patches to support [multiple modes](https://bottest.wiki.kernel.org/coccicheck): report(, context, org,) and patch.  Adding that support requires at least two versions of every patch rule, because the metadecl section is structurally different between the report and patch (and context) forms.  It also has to be tested for all supported modes.

If the replacement were something likely to be done repeatedly over time, as with patches for consistent style, the overhead is worthwhile.

But this is a one-shot tree-wide transformation, that at most might be used again for new PRs that were developed using the existing structures (and which won't build once the structure changes).

I'm not sure this is worth the time.  However, we could restructure `scripts/coccinelle` to distinguish between patches that are expected to be run repeatedly, and those that are single-use but serve as examples for future related transformations.

**Review by pabigot — APPROVED** [review]

**Comment by tbursztyka** [comment]
@pabigot thanks for the details, don't bother yes.

**Review by jfischer-no — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by tbursztyka** [comment]
@carlescufi I guess that's ready to be merged any time you want. (I'll rebase if new commits makes it non-mergable"

**Comment by pabigot** [comment]
@tbursztyka @carlescufi The merge of #17631 requires an update to this.

https://github.com/pabigot/zephyr/commits/nordic/pr/23407 has the necessary fixup patch, plus a second DNM patch that has the coccinelle script that will fix other cases if more things get merged before this gets merged.

**Comment by carlescufi** [comment]
> @tbursztyka @carlescufi The merge of #17631 requires an update to this.
> 
> https://github.com/pabigot/zephyr/commits/nordic/pr/23407 has the necessary fixup patch, plus a second DNM patch that has the coccinelle script that will fix other cases if more things get merged before this gets merged.

Thanks @pabigot. I believe @tbursztyka is off on holiday today, so would it be possible to get this PR merged and then merge the PR (I think you listed the wrong one since you point at this very one) that has the required fixes?

**Comment by tbursztyka** [comment]
@carlescufi Yes it's off here, but I had to login, so I can do it

**Comment by pabigot** [comment]
@tbursztyka Three more patches are necessary due to problems with `Z_SYSCALL_SPECIFIC_DRIVER` and subsequent changes in master.  These are in https://github.com/pabigot/zephyr/commits/nordic/pr/23407

I can apply them and similar patches resulting from further delays if you authorize me to push to your branch---I don't do that without clear prior assent.

File: include/syscall_handler.h
Comment (pabigot) [inline include/syscall_handler.h:458]: Missed the rename of `_init_fn` to `_api` to match use in the call.

**Review by pabigot — CHANGES_REQUESTED** [review]

**Comment by tbursztyka** [comment]
@pabigot can you re-check?

**Comment by pabigot** [comment]
@tbursztyka Yes, this one seems to have everything: maxim test of syscall driver works, no additional changes snuck in.

**Review by pabigot — APPROVED** [review]

**Pull request merged**
